HWY-35: fork choice

afck opened the pull request:
### Overview
This implements the original GHOST rule: Only every honest validator's latest vote is counted, and if it doesn't vote for a child of the previously selected block anymore, it is discarded, and _not_ replaced with its sender's most recent vote for a child.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/HWY-35

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [x] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](https://drone-auto.casperlabs.io/) system. This is necessary to run tests on this PR.
- [ ] Do not forget to run `bors r+` if GitHub policy is not enforced, e.g. when merging into another feature branch. It may be omitted under some circumstances if this PR intentionally assumes that integration tests will fail but will be fixed with the future PRs.

goral09 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,159 @@
+            .and_modify(|tally| tally.add(bhash, weight))
+            .or_insert_with(|| Tally::new(bhash, weight));
+    }
+}
```
I would like to see some unit tests for `Tally` and `Tallies`.

goral09 replied:
`find_decided` and `filter` look important.

goral09 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -186,10 +190,29 @@ impl<C: Context> State<C> {
+            let bhash = &self.vote(obs.correct()?).block;
+            Some((self.block(bhash).height, bhash, *w))
+        };
+        let mut tallies: Tallies<C> = pan.iter().zip(&self.weights).filter_map(to_entry).collect();
```
So the assumption is that `pan` and `weights` are ordered the same way? Who enforces that?

afck replied:
Yes, internally `highway-core` uses indices to identify validators. That's much faster and wastes less space than putting everything in maps by validator-public-key or something.

It will also be the assumption later, when we use Merkle trees to store the panoramas.

We enforce it by checking panoramas' length and consistency, and eventually we'll add signature checking: A message sent by validator `i` must be signed by validator `i`.

goral09 replied:
As a side not: sometimes I know the answers to my questions but I ask them to force an author to think about his solution.

Good that you remember it :+1:

goral09 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    /// Removes all votes for blocks that are not descendants of `bhash`.
+    pub fn filter(self, height: u64, bhash: &'a C::VoteHash, state: &'a State<C>) -> Self {
```
It's not really a `filter` method since it doesn't accept the predicate but has the logic hardcoded. Can you change the method name to reflect what it's doing?

goral09 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -207,6 +230,25 @@ impl<C: Context> State<C> {
         self.find_in_swimlane(&vote.skip_idx[i], seq_number)
     }
 
+    pub fn find_ancestor<'a>(
```
Can you add a comment to this method?

goral09 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -207,6 +230,25 @@ impl<C: Context> State<C> {
+        }
+        let diff = block.height - height;
+        // We want to make the greatest step 2^i such that 2^i <= diff.
+        let max_i = (diff + 1).next_power_of_two().trailing_zeros() as usize - 1;
```
You have the utility method for this one now.

goral09 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -207,6 +230,25 @@ impl<C: Context> State<C> {
+        if block.height < height {
+            return None;
+        }
+        if block.height == height {
+            return Some(hash);
+        }
+        let diff = block.height - height;
+        // We want to make the greatest step 2^i such that 2^i <= diff.
+        let max_i = (diff + 1).next_power_of_two().trailing_zeros() as usize - 1;
+        let i = max_i.min(block.skip_idx.len() - 1);
+        self.find_ancestor(&block.skip_idx[i], height)
```
Is it worth extracting this bit into an utility method? You're doing the same thing in 3 places now (parents skip list, swimlane skip list and now this.

afck replied:
But now it's only two simple lines (and the recursive call afterwards):
```rust
let max_i = log2(diff) as usize;
let i = max_i.min(block.skip_idx.len() - 1);
```
Also, it's only two places, isn't it? This _is_ the parents/ancestors lookup, and the other one is the swimlane one.

goral09 replied:
Hmm, yeah. I misremembered.

goral09 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -44,6 +45,8 @@ impl<C: Context> WireVote<C> {
 #[derive(Debug)]
 pub struct State<C: Context> {
+    /// The validator's voting weights.
+    weights: Vec<u64>,
```
In the current EE weights are u512.

afck replied:
I know, but they're used in the fork choice computation so much that it would really slow us down, and a granularity of 2<sup>64</sup> is more than enough for vote weights. We just need to scale the `U512`s down proportionally when creating the weight list.

goral09 replied:
Ah, right. I forgot. So it will be the responsibility of an era supervisor (or the function that instantiates the consensus protocol instance) to scale it.

goral09 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,159 @@
+#[derive(Clone)]
+pub struct Tally<'a, C: Context> {
+    /// The block with the highest weight, and the highest hash if there's a tie.
+    max: (u64, &'a C::VoteHash),
```
Is u64 a height of weight?

afck replied:
A weight, since each `Tally` only refers to a single height anyway.
I also noticed there's a danger of confusing heights and weights here… :confused:
Not sure if it's worth adding separate types?
```rust
#[derive(…)]
struct Weight(u64);
```
Or just aliases?
```rust
type Weight = u64;
```
Maybe I'll try the former…

goral09 replied:
A separate type is safer but maybe type alias will suffice here since like you're saying a Tally refers to a single height anyway. 

Whatever you decide.

goral09 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,159 @@
+
+impl<'a, 'b, C: Context> IntoIterator for &'b Tally<'a, C> {
+    type Item = (&'a C::VoteHash, u64);
+    type IntoIter = Box<dyn Iterator<Item = Self::Item> + 'b>;
```
Can you explain to a Rust noob why `Box<dyn Iterator>` is necessary?

afck replied:
I wish there was a _good_ reason, but unfortunately, as far as I know, the following is still true:
The iterator created by my implementation, if mentioned by the compiler, is something like
```rust
iter::Map<btree_map::Iter<'_, &C::VoteHash, u64>, [closure@tallies.rs:32]>
```
and there's still no way to express the `[closure…]` part in actual Rust code. So it has a type that you can't spell out! :weary: 
A while ago, the `impl Iterator<(&C::VoteHash, u64)>` syntax for return types was added as a partial solution to that. It basically tells the compiler: "You know exactly _which_ `Iterator` implementation I'm talking about anyway — the one returned by the function body. So use that." You still can't express `[closure…]`, but now you don't have to.
Unfortunately that syntax can't be used in associated types yet. So I have to wrap it in a `Box` and use dynamic dispatch here, just so that I can express the associated type! :pensive:

goral09 replied:
Thank you for this detailed explanation.

goral09 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,159 @@
+        let map_compatible =
+            |(h, t): (u64, Tally<'a, C>)| t.filter(height, bhash, state).map(|t| (h, t));
+        // All tallies at `height` and lower can be removed, too.
+        let relevant_heights = self.0.into_iter().rev().take_while(|(h, _)| *h > height);
```
Do you need `self.0` if you derive `Deref` and `DerefMut`?

goral09 replied:
Does iterator guarantee you any deterministic order? I'm talking about `into_iter` and `take_while`.

afck replied:
Unfortunately yes: `Deref` gives me a `&BTreeMap`, and `DerefMut` a `&mut BTreeMap`. But `into_iter` _consumes_ the map, i.e. it needs a `BTreeMap`.

afck replied:
Yes, `into_iter` on `BTreeMap`s iterates in the order of the keys, `rev` reverses that order, and `take_while` doesn't change the order.

goral09 replied:
Cool, I wanted to make sure that `take_while` is not called on unordered collection. I should have known better.

afck commented:
I addressed the comments, and added a `Weight` type (not just alias). It adds quite a bit of code, though, so I'm not sure whether we should also do it for `Height`. (And `SequenceNumber`?)
But I guess if we do so, that can be a separate PR, because the `Weight` type resolves the concerns in this one at least, preventing height and weight from getting mixed up.

fizyk20 commented on execution-engine/consensus/highway-core/src/state.rs:
```diff
@@ -335,9 +394,15 @@ mod tests {
-        fn val(mut self, values: Vec<&'static str>) -> Self {
-            self.values = Some(values);
+        /// Adds a value to the vote, turning it into a new block.
+        pub fn val(mut self, value: &'static str) -> Self {
```
Should this function's name reflect the fact that it mutates the vote? Since it returns `Self`, something like `with_value`, maybe?

afck replied:
Done.

afck commented:
> I'm sorry for making you do this!

I'm always in favor of type safety! :+1:

fizyk20 commented on execution-engine/consensus/highway-core/src/tallies.rs:
```diff
@@ -0,0 +1,251 @@
+            if let Some(tally) = self.get(&height) {
+                h_tally.extend(tally);
+            }
+            // If any block received more than 50%, a decision can be made: Either that block is
+            // the fork choice, or we can pick its highest scoring child from `prev_tally`.
+            if h_tally.max_w() * 2 > total_weight {
+                return Some(
+                    match prev_tally.filter_descendants(height, h_tally.max_bhash(), state) {
+                        Some(filtered) => (height + 1, filtered.max_bhash()),
+                        None => (height, h_tally.max_bhash()),
+                    },
+                );
+            }
```
This slightly confused me. We will be filtering the descendants of the block returned from here in `fork_choice` anyway, right? Why not just return the block with more than 50% from here, then? As it is now, it looks like we're trying to do a part of `fork_choice`'s work here in advance, which makes it less clear what the purpose of `find_decided` is. At least to me, maybe I'm missing something here?

afck replied:
That's what I wanted to do at first, too.
However, it can happen that the first height at which `h_tally.max_w() * 2 > total_weight` is the one that we've already decided in the last round. E.g. at height 2 there's `a2` and everyone is voting for that. But at height 3 there are `a3`, `b3`, `c3`, with weights 3, 4 and 5 (no majority). So we'd select `a2` again, the filter step would remove nothing and we would loop forever.

The obvious solution is to just pass in the `height` we selected last time, and make sure we stop iterating at `height + 1`, even if we have to take a plurality instead of majority.

However, I thought I'd do something "elegant" (or, in your, more accurate, words: slightly confusing), and make sure we _always_ make a decision at at least one level higher than the highest level with a strict majority… :grimacing:

fizyk20 replied:
Ah, I see! As it is now, we also call `find_decided` in a loop, which adds the dropped parents to the tally _again_ even if we drop them via filtering - do I understand correctly?

What I was thinking was that after we find the block with majority, we'll start to only move up. So we'd call `find_decided` once, and then recursively: filter to only descendants of the block with maximum weight and choose the block with maximum weight among those left with the lowest height. So in every iteration of filtering, we'd move one level up, and `find_decided` would only serve as a way of finding the starting point for climbing, so to speak.

But of course it's possible that I still don't get some nuance that makes my idea not viable...

afck replied:
> adds the dropped parents to the tally _again_ even if we drop them via filtering

Yes, `find_decided` doesn't modify `self`, so the elements that were dropped in line 140 are still there. However, everything that isn't a (strict) descendant of the block returned by `find_decided` will be removed in the `Tallies::filter_descendants` call, so the size of `tallies` does strictly decrease in each iteration.

I agree that your algorithm would work, too, and it would probably be fine. However, if the first attempt to find decided ends up returning something with a very low height, we'd go through each height above that explicitly.
E.g. let's say Alice votes for `a0` (height 0), and Bob and Carol vote for `b99`, which is a descendant of `b0`.
My implementation will prune Alice's vote in the first step, and jump directly to height 99 in the second step. If we'd only use `find_decided` as a starting point, we'd probably go through 99 iterations here.

fizyk20 replied:
I see. Alright, that makes sense :+1:

afck commented:
bors r+

afck commented:
bors r+
